frontend/src/Contexts/productContext.js

```diff
@@ -0,0 +1,47 @@
+import axios from "axios";
+import { createContext, useContext, useEffect, useReducer } from "react";
+import { productReducer, initialProductData } from "../Reducers/productReducer";
+
+const ProductContext = createContext({
+  state: initialProductData,
+  dispatch: () => {},
+});
+
+const ProductProvider = ({ children }) => {
+  const [state, dispatch] = useReducer(productReducer, initialProductData);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const { data } = await axios.get("/api/products");
+        dispatch({
+          type: "INITIALIZE_DATA",
+          payload: data,
+        });
+      } catch (error) {
+        console.log(error);
+      }
+    };
+
+    fetchData();
+  }, []);
+
+  const value = {
+    dispatch: dispatch,
+    categoriesList: state.categoriesList,
+    productsList: state.productsList,
+    sortByRating: state.sortByRating,
+    priceRange: state.priceRange,
+    sliderValue: state.sliderValue,
```

Comment:
Here instead of doing like this you can do something this,
```javascript
const value = {
dispatch,
...state,
sortByLowAndHigh: state.sortByHighLow,
}
```

This will decrease the number of lines to type